number_parser/parser.py

```diff
@@ -153,41 +153,61 @@ def _normalize_tokens(token_list):
     return [_strip_accents(token.lower()) for token in token_list]
 
 
-def _normalize_dict(lang_dict):
+def _normalize_dict(lang_data):
     """Removes the accent from each key of input dictionary"""
-    return {_strip_accents(word): number for word, number in lang_dict.items()}
+    return {_strip_accents(word): number for word, number in lang_data.items()}
 
 
-def _is_cardinal_token(token, lang_dict):
-    if token in lang_dict.all_numbers:
+def _is_cardinal_token(token, lang_data):
+    """Checks if the given token is a cardinal number and returns token"""
+    if token in lang_data.all_numbers:
         return token
     return None
 
 
-def _is_ordinal_token(token, lang_dict):
-    token = _apply_cardinal_conversion(token, lang_dict)
-    return _is_cardinal_token(token, lang_dict)
+def _is_ordinal_token(token, lang_data):
+    """Checks if the given token is a ordinal number and returns token"""
+    if _is_cardinal_token(token, lang_data) is None:
+        return _is_number_token(token, lang_data)
+    return None
 
 
-def _is_number_token(token, lang_dict):
-    return _is_cardinal_token(token, lang_dict) or _is_ordinal_token(token, lang_dict)
+def _is_number_token(token, lang_data):
+    """
+    Checks if the given token belongs to either cardinal or ordinal numbers
+    and returns the cardinal form.
+    """
+    token = _apply_cardinal_conversion(token, lang_data)
+    return _is_cardinal_token(token, lang_data)
 
 
-def _apply_cardinal_conversion(input_string, language):
-    # this will be coming from the LanguageData
+def _apply_cardinal_conversion(token, lang_data):
+    """Converts ordinal tokens to cardinal while leaving other tokens unchanged."""
     CARDINAL_DIRECT_NUMBERS = {'first': 'one', 'second': 'two', 'third': 'three', 'fifth': 'five', 'eighth': 'eight',
                                'ninth': 'nine', 'twelfth': 'twelve'}
-    input_string = input_string.lower()
+
     for word, number in CARDINAL_DIRECT_NUMBERS.items():
-        input_string = input_string.replace(word, number)
-    input_string = re.sub(r'ieth$', 'y', input_string)
-    input_string = re.sub(r'th$', '', input_string)
-    return input_string
+        token = token.replace(word, number)
+
+    token_cardinal_form_1 = re.sub(r'ieth$', 'y', token)
+    if _is_cardinal_token(token_cardinal_form_1, lang_data) is not None:
+        return token_cardinal_form_1
+
+    token_cardinal_form_2 = re.sub(r'th$', '', token)
+    if _is_cardinal_token(token_cardinal_form_2, lang_data) is not None:
+        return token_cardinal_form_2
+
+    return token
 
 
 def parse_ordinal(input_string, language='en'):
```

Comment:
This now has acts as the third interface after ```parse , parse_number ```. All these interfaces have the ```language``` parameter while the rest use the ```lang_data```.
This currently converts all tokens to their cardinal form and then gives it to parse_number , which means that using ```parse_ordinal``` with only cardinal , cardinal/ordinal mixture and multiple ordinals all work , which is not completely accurate. (added these cases in tests)